include/dlaf/common/utils.h

```diff
@@ -12,6 +12,7 @@
 
 /// @file
 
+#include <iostream>
 #include <sstream>
 #include <string>
```

Comment:
I think you can remove all headers now.

I see you decided to move `concat` to the `assert.h` header. I completely agree that it should be in an `internal` namespace since it is something we don't want to provide in the public API of this library. About the include file, I don't think it is worth having an additional file just for this macro with related data structure, so the the options I see are:
- move also `SOURCE_LOCATION` to the `assert.h` header file and remove the `common/utils.h`
- move back `concat` to this header file (but in the `internal` namespace as you correctly did)

I would go for the first one, if there is not any other side effect, because as you modified the `concat` function it is likley it won't be used anywhere else (considering that it concatenates using a newline).

Reply:
> move also SOURCE_LOCATION to the assert.h header file and remove the common/utils.h

`SOURCE_LOCATION` is used in `communication/error.h` and `cuda/error.h` (also `cublas/error.h` in the future), perahps I should rename the header to `location.h` instead of `utils.h` and just keep it there?

Reply:
Good point! I like your proposal of renaming it to `location.h`.